### PR TITLE
BUG query self indexing and sorting

### LINKED ISSUE #13: bug in indexing
This line is wrong: https://github.com/esheldon/smatch/blob/master/smatch/matcher.py#L270

### LINKED ISSUE #14: idx2 is sorted wrongly
The query_self method sorts idx 2 wrong I think.

### PR DESCRIPTION
**Pull request by beckermr**
The query_self method had some odd bugs that this PR fixes.

closes #13 
closes #14

### CODE REVIEW

**Comment by esheldon** [comment]
should there be a new unit test checking this bug?

this is also an opportunity to add pytest like tests

**Comment by beckermr** [comment]
I don't want to spend the time on testing right now. Also the api here is not quite right and so that should be fixed first. I'd prefer to merge and come back to fix later.

**Comment by esheldon** [comment]
OK, but then we should not make a release yet

**Comment by beckermr** [comment]
A patch release now has more net value in fixing a serious bug. I don't see why we wait.

**Comment by esheldon** [comment]
Without a test how do we know if this is an improvement over the old code?

**Comment by beckermr** [comment]
Half of the lines are copied from above and the new lines I've run locally. Happy to wait.

**Comment by beckermr** [comment]
@esheldon This is ready for review. cc @erykoff

File: smatch/matcher.py [hunk 1]
```diff
@@ -260,28 +256,20 @@ def query_self(self, radius, min_match=1, eps=0.0, return_indices=False):
...
         idx = self.tree.query_ball_tree(self.tree, angle, eps=eps)
-        # The matching works best when sorting with the most matches first
-        len_arr = np.array([len(j) for j in idx])
-        st = np.argsort(len_arr)[:: -1]
-        match_id = np.full(len(idx), -1, dtype=np.int32)
-        idx2 = []
-        for j in st:
-            if match_id[j] < 0 and len_arr[j] >= min_match:
-                match_id[idx[j]] = j
-                idx2.append(idx[j])
```
Comment (erykoff) [inline smatch/matcher.py:271]: Why did you get rid of this section?  It's absolutely critical to remove dependence on ordering of the input catalog.

File: smatch/matcher.py [hunk 2]
```diff
@@ -226,17 +226,13 @@ def query_radius(self, lon, lat, radius, eps=0.0, return_indices=False):
...
         """Match the list of lon/lat to itself.
 
         Parameters
         ----------
         radius : `float`
             The match radius in degrees.
-        min_match : `int`, optional
-            Minimum number of matches to count as a match.
-            If min_match=1 then all positions will be returned since every
-            position will match at least to itself.
```
Comment (erykoff) [inline smatch/matcher.py]: I depend on this functionality.

File: smatch/matcher.py (hunk 1 above)
Comment (beckermr) [reply smatch/matcher.py:271]: The ordering relative to the inputs changing is a bug.

File: smatch/matcher.py (hunk 2 above)
Comment (beckermr) [reply smatch/matcher.py]: arg ok.

File: smatch/matcher.py (hunk 1 above)
Comment (beckermr) [reply smatch/matcher.py:271]: How does one interpret the list of lists if the ordering has changed relative to the inputs?

File: smatch/tests/test_matcher.py [hunk 3]
```diff
@@ -204,3 +206,76 @@ def test_sphdist():
...
+    idx = mch.query_self(6e4/3600)
+    assert any(len(_idx) > 1 for _idx in idx)
+
+    for ic in range(ra.shape[0]):
+        idxc = []
+        for ip in range(ra.shape[0]):
+            sep = sphdist(ra[ic], dec[ic], ra[ip], dec[ip])
+            if sep < 6e4/3600:
+                idxc.append(ip)
+        assert set(idxc) == set(idx[ic])
```
Comment (beckermr) [inline smatch/tests/test_matcher.py:224]: @erykoff I checked explicitly and the old code fails this simple test and so is wrong.

File: smatch/matcher.py (hunk 2 above)
Comment (beckermr) [reply smatch/matcher.py]: I put this back and wrote a test for it.

**Review by erykoff — APPROVED** [review]
I'm happy with this now, and rubinenv is protected from the update as well so I can transition to the new version on a timescale that won't break anything I use.

**Comment by beckermr** [comment]
That awesome. You might add a test for the effects of duplicate points in the code for groups. Up to you.

**Comment by beckermr** [comment]
@esheldon Eli is happy and the tests pass. Any other comments here?

**Comment by esheldon** [comment]
If you guys are happy with it then I'm ok

**Review by esheldon — APPROVED** [review]
